doc/design/logging.md

```diff
@@ -0,0 +1,385 @@
+# Logging
+
+## Summary and problem description
+
+Logging is a crucial part of a framework since it informs the developer and
+user of the framework about anomalies and helps with debugging by providing
+contextual information.
+
+Additionally, when integrated into a separate framework, the log messages should
+be forwarded in order to have a single logging infrastructure.
+
+The logging should also be performant and not allocate memory. Ideally, it should
+be possible to disable it at compile time and let the compiler optimize it away.
+
+It should also be possible to alter the behavior of the logging via environment
+variables, e.g. the log level.
+
+## Terminology
+
+### Log levels and how to use them
+
+| LogLevel | Usage |
+|----------|-------|
+| FATAL    | For fatal and non-recoverable errors which essentially lead to termination of the program. |
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens which is not considered a severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
```

Comment:
I think this does not fit. Here a short example, `storable_function.inl:113`. The copy assignment could then log:
```
trace:: this operations is needed for destroy, then changed to source (rhs) operations
```

this would be only weird and confusing.

But when I think about trace, the word itself, then I would like to trace steps etc. So I would put as recommendation or guideline here that lifetime changes (creation/destruction) of vital objects could be logged with trace as well as important achievements in the lifetime of an application. When it is a little nebulous here an example.

* inform the user about any shared memory object that was created and mapped into process space
* when roudi is started up correctly
* whenever a waitset or listener is created or the underlying condition variables
* new applications which register or unregister at roudi

What do you think?

Reply:
log4j uses the following definition `Designates finer-grained informational events than the DEBUG`. So how about `Finer-grained information than DEBUG` to keep it short?

Reply:
We can do this. In the last days I tried to figure out what and where should be printed in which log level. I came up with the following definitions:

* **Trace** - Lifetime changes of vital constructs or track important intermediate steps
  * example: creation/remove of shared memory segments; log that roudi is ready to rumble :smile: 
* **Debug** - Like info but less relevant for the user.
  * example: use default config file, set certain log level
* **Info** - All log outputs which could be relevant for the daily user

I think the remaining log levels are clear.

But to keep things short, I agree with your definition!

Reply:
@elfenpiff I would say `DEBUG` has no relevance to the user but only the developer.

How about `Anything which could be relevant for the daily user` for `INFO`?

I also wouldn't narrow it down too much. The transitions are fluid and depending on the context a lifetime change could be a `DEBUG` in a case where it is critical and a `TRACE` in another case.

Reply:
> How about Anything which could be relevant for the daily user for INFO?

I do not want to delay this PR too much but I find this somehow interesting to nail down when to debug and when to trace since I have no idea when to do what.

But in the end one could say: Debug is the Info log level for the fellow developer.

Reply:
@elfenpiff I like the suggestion for debug :)

We would have
| LogLevel       | Definition |
|------------|------------|
| INFO | Anything which could be relevant for the daily user     |
| DEBUG  | Anything that is helpful for debugging, i.e. the info log level for the fellow developer     |
| TRACE  | Anything that might be helpful for debugging but would be too verbose in most cases |